src/main/java/pvpmode/internal/server/PvPDataManager.java

```diff
@@ -0,0 +1,174 @@
+package pvpmode.internal.server;
+
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.*;
+
+import net.minecraft.entity.player.*;
+import net.minecraft.nbt.*;
+import pvpmode.api.common.utils.PvPCommonUtils;
+import pvpmode.api.server.PvPData;
+import pvpmode.api.server.utils.PvPServerUtils;
+
+public class PvPDataManager
+{
+
+    private final ServerProxy server;
+    private Path playerDataRootDir;
+
+    private final Map<UUID, PvPDataImpl> playerData = new HashMap<> ();
+
+    public PvPDataManager (ServerProxy server)
+    {
+        this.server = server;
+    }
+
+    public void onServerStarting () throws IOException
+    {
+        playerDataRootDir = server.getWorldDataFolder ().resolve ("playerdata");
+
+        Files.createDirectories (playerDataRootDir);
+    }
+
+    public void onLogin (EntityPlayer player)
+    {
+        loadPlayerData (player.getUniqueID (), compatibilityPatch (player));
+        setCacheData (player);
+    }
+
+    public void onLogout (EntityPlayer player)
+    {
+        setCacheData (player);
+    }
+
+    private void setCacheData (EntityPlayer player)
+    {
+        PvPDataImpl data = (PvPDataImpl) getPlayerData (player);
+
+        data.setCanFlyCache (PvPServerUtils.canFly (player));
+        data.setCreativeCache (PvPCommonUtils.isCreativeMode (player));
+    }
+
+    public PvPData getPlayerData (EntityPlayer player)
+    {
+        return getPlayerData (player.getUniqueID ());
+    }
+
+    public PvPData getPlayerData (UUID playerUUID)
+    {
+        if (!playerData.containsKey (playerUUID))
+        {
+            playerData.put (playerUUID, this.loadPlayerData (playerUUID, null));
+        }
+        return playerData.get (playerUUID);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void saveAndClearData ()
+    {
+        playerData.forEach (this::savePlayerData);
+        Collection<UUID> presentPlayers = new HashSet<> ();
+        ((Collection<EntityPlayerMP>) server.getServerConfigurationManager ().playerEntityList)
+            .forEach (player -> presentPlayers.add (player.getUniqueID ()));
```

Comment:
```suggestion
        Collection<UUID> presentPlayers = ((Collection<EntityPlayerMP>) server.getServerConfigurationManager ().playerEntityList)
                .stream ().map (Entity::getUniqueID).collect (Collectors.toSet())
```

Reply:
Additionally, it might be a bit clearer to just convert everything to a `List` since everything is actually a list.
```suggestion
        List<UUID> presentPlayers = ((List<EntityPlayerMP>) server.getServerConfigurationManager ().playerEntityList)
                .stream ().map (Entity::getUniqueID).collect (Collectors.toList())
```

Reply:
Regarding the Stream API: Good point.
I use the `Collection` interface usually if it doesn't matter if it's a list or set or something else, and see no added value here for using the list interface.